File: app/src/main/java/com/adasa/foodplan/domain/usecase/GetShoppingListUseCase.kt
package com.adasa.foodplan.domain.usecase

import com.adasa.foodplan.data.repository.IngredientRepository
import com.adasa.foodplan.data.repository.MealPlanRepository
import com.adasa.foodplan.data.repository.RecipeRepository
import com.adasa.foodplan.domain.model.Recipe
import com.adasa.foodplan.domain.model.RecipeContribution
import com.adasa.foodplan.domain.model.SelectableRecipe
import com.adasa.foodplan.domain.model.ShoppingCategory
import com.adasa.foodplan.domain.model.ShoppingItem
import com.adasa.foodplan.domain.model.ShoppingList
import com.adasa.foodplan.domain.model.ShoppingPeriod
import com.adasa.foodplan.domain.model.ShoppingUnit
import kotlinx.datetime.LocalDate
import javax.inject.Inject

class GetShoppingListUseCase @Inject constructor(
    private val mealPlanRepository: MealPlanRepository,
    private val recipeRepository:   RecipeRepository,
    private val ingredientRepository: IngredientRepository
) {
    /**
     * Build a shopping list for [startDate]..[endDate].
     *
     * @param selectedRecipeIds When non-null, only ingredients from these recipe IDs
     *   are included in the list. The full recipe list in [ShoppingPeriod] always
     *   contains all recipes so the UI can show the selector.
     *   null means all recipes are included.
     */
    suspend operator fun invoke(
        startDate:         LocalDate,
        endDate:           LocalDate,
        selectedRecipeIds: Set<String>? = null
    ): ShoppingList {

        // 1. Fetch all day plans in the range
        val dayPlans = mealPlanRepository.getDayPlansForRange(startDate, endDate)

        // 2. All unique recipe IDs in the period
        val allRecipeIds = dayPlans.values
            .flatMap { it.meals }
            .map { it.recipeId }
            .distinct()

        // 3. Fetch all recipes (full list for the period card selector)
        val allRecipes = allRecipeIds.mapNotNull { id ->
            recipeRepository.getRecipeWithIngredients(id)
        }

        // 4. Apply optional recipe filter
        val activeRecipeIds = selectedRecipeIds ?: allRecipeIds.toSet()
        val activeRecipes   = allRecipes.filter { it.id in activeRecipeIds }

        // 5. Build ingredient + sub-recipe caches
        val allIngredientIds = collectAllIngredientIds(allRecipes)
        val ingredientCache  = allIngredientIds.mapNotNull { id ->
            ingredientRepository.getIngredientById(id)
        }.associateBy { it.id }

        val subRecipeIds = allRecipes
            .flatMap { it.ingredients.mapNotNull { ri -> ri.subRecipeId } }
            .distinct()
        val subRecipeCache = subRecipeIds.mapNotNull { id ->
            recipeRepository.getRecipeWithIngredients(id)
        }.associateBy { it.id }

        // 6. Aggregate: ingredientId -> Map<recipeId, Pair<recipeName, grams>>
        //    Tracking per-recipe so we can show the breakdown on each item row.
        val aggregated = mutableMapOf<String, MutableMap<String, Pair<String, Double>>>()

        activeRecipes.forEach { recipe ->
            val portionCount = dayPlans.values
                .flatMap { it.meals }
                .count { it.recipeId == recipe.id }
                .toDouble().coerceAtLeast(1.0)

            expandRecipeIngredients(
                recipe            = recipe,
                portionMultiplier = portionCount,
                subRecipeCache    = subRecipeCache,
                recipeId          = recipe.id,
                recipeName        = recipe.name
            ) { ingredientId, grams, recipeId, recipeName ->
                val byRecipe = aggregated.getOrPut(ingredientId) { mutableMapOf() }
                val existing = byRecipe[recipeId]
                byRecipe[recipeId] = recipeName to ((existing?.second ?: 0.0) + grams)
            }
        }

        // 7. Build ShoppingItems with per-recipe contributions
        val items = aggregated.mapNotNull { (ingredientId, byRecipe) ->
            val ingredient    = ingredientCache[ingredientId] ?: return@mapNotNull null
            val contributions = byRecipe.entries
                .map { (recipeId, pair) -> RecipeContribution(recipeId, pair.first, pair.second) }
                .sortedBy { it.recipeName }
            ShoppingItem(
                ingredientId  = ingredientId,
                name          = ingredient.name,
                totalGrams    = contributions.sumOf { it.grams },
                unit          = ShoppingUnit.GRAMS,
                contributions = contributions
            )
        }

        // 8. Group by category
        val categories = items
            .groupBy { ingredientCache[it.ingredientId]?.category }
            .map { (category, catItems) ->
                ShoppingCategory(
                    name  = category?.displayName ?: "Other",
                    emoji = category?.emoji ?: "🛒",
                    items = catItems.sortedBy { it.name }
                )
            }
            .sortedBy { it.name }

        return ShoppingList(
            period = ShoppingPeriod(
                startDate = startDate,
                endDate   = endDate,
                recipes   = allRecipes
                    .map { SelectableRecipe(it.id, it.name) }
                    .distinctBy { it.id }
                    .sortedBy { it.name }
            ),
            categories = categories
        )
    }

    private fun collectAllIngredientIds(
        recipes:        List<Recipe>,
        subRecipeCache: Map<String, Recipe> = emptyMap()
    ): Set<String> {
        val ids = mutableSetOf<String>()
        recipes.forEach { recipe ->
            recipe.ingredients.forEach { ri ->
                when {
                    ri.ingredientId != null -> ids.add(ri.ingredientId)
                    ri.subRecipeId  != null -> {
                        val sub = subRecipeCache[ri.subRecipeId]
                        if (sub != null) ids.addAll(collectAllIngredientIds(listOf(sub), subRecipeCache))
                    }
                }
            }
        }
        return ids
    }

    private fun expandRecipeIngredients(
        recipe:            Recipe,
        portionMultiplier: Double,
        subRecipeCache:    Map<String, Recipe>,
        recipeId:          String,
        recipeName:        String,
        visited:           MutableSet<String> = mutableSetOf(),
        onIngredient:      (ingredientId: String, grams: Double, recipeId: String, recipeName: String) -> Unit
    ) {
        if (recipe.id in visited) return
        visited.add(recipe.id)

        recipe.ingredients.forEach { ri ->
            when {
                ri.ingredientId != null && ri.grams != null -> {
                    onIngredient(ri.ingredientId, ri.grams * portionMultiplier, recipeId, recipeName)
                }
                ri.subRecipeId != null && ri.portions != null -> {
                    val sub = subRecipeCache[ri.subRecipeId] ?: return@forEach
                    expandRecipeIngredients(
                        recipe            = sub,
                        portionMultiplier = portionMultiplier * ri.portions,
                        subRecipeCache    = subRecipeCache,
                        recipeId          = recipeId,
                        recipeName        = recipeName,
                        visited           = visited,
                        onIngredient      = onIngredient
                    )
                }
            }
        }
    }
}

File: flutter-app/lib/use_case/get_shopping_list_use_case.dart
import 'package:foodplan/data/repository/ingredient_repository.dart';
import 'package:foodplan/data/repository/meal_plan_repository.dart';
import 'package:foodplan/data/repository/recipe_repository.dart';
import 'package:foodplan/domain/model/ingredient.dart';
import 'package:foodplan/domain/model/recipe.dart';
import 'package:foodplan/domain/model/shopping_list.dart';

/// Mirrors the Kotlin GetShoppingListUseCase.
///
/// Builds a [ShoppingList] for [startDate]..[endDate], optionally filtered
/// to [selectedRecipeIds].  null means all recipes are included.
class GetShoppingListUseCase {
  GetShoppingListUseCase({
    required MealPlanRepository mealPlanRepository,
    required RecipeRepository recipeRepository,
    required IngredientRepository ingredientRepository,
  })  : _mealPlanRepository = mealPlanRepository,
        _recipeRepository = recipeRepository,
        _ingredientRepository = ingredientRepository;

  final MealPlanRepository _mealPlanRepository;
  final RecipeRepository _recipeRepository;
  final IngredientRepository _ingredientRepository;

  Future<ShoppingList> call(
    DateTime startDate,
    DateTime endDate, {
    Set<String>? selectedRecipeIds,
  }) async {
    // 1. Fetch all day plans in the range.
    final dayPlans =
        await _mealPlanRepository.getDayPlansForRange(startDate, endDate);

    // 2. All unique recipe IDs in the period.
    final allRecipeIds = dayPlans.values
        .expand((dp) => dp.meals)
        .map((s) => s.recipeId)
        .toSet()
        .toList();

    // 3. Fetch all recipes (full list for the period card selector).
    final allRecipes = <Recipe>[];
    for (final id in allRecipeIds) {
      final recipe = await _recipeRepository.getRecipeWithIngredients(id);
      if (recipe != null) allRecipes.add(recipe);
    }

    // 4. Apply optional recipe filter.
    final activeRecipeIds = selectedRecipeIds ?? allRecipeIds.toSet();
    final activeRecipes =
        allRecipes.where((r) => activeRecipeIds.contains(r.id)).toList();

    // 5. Build ingredient + sub-recipe caches.
    final allIngredientIds =
        _collectAllIngredientIds(allRecipes, {});
    final ingredientCache = <String, Ingredient>{};
    for (final id in allIngredientIds) {
      final ing = await _ingredientRepository.getIngredientById(id);
      if (ing != null) ingredientCache[id] = ing;
    }

    final subRecipeIds = allRecipes
        .expand((r) => r.ingredients)
        .map((ri) => ri.subRecipeId)
        .whereType<String>()
        .toSet();
    final subRecipeCache = <String, Recipe>{};
    for (final id in subRecipeIds) {
      final recipe = await _recipeRepository.getRecipeWithIngredients(id);
      if (recipe != null) subRecipeCache[id] = recipe;
    }

    // 6. Aggregate: ingredientId -> Map<recipeName, grams>
    final aggregated = <String, Map<String, double>>{};

    for (final recipe in activeRecipes) {
      final portionCount = dayPlans.values
          .expand((dp) => dp.meals)
          .where((s) => s.recipeId == recipe.id)
          .length
          .toDouble()
          .clamp(1.0, double.infinity);

      _expandRecipeIngredients(
        recipe: recipe,
        portionMultiplier: portionCount,
        subRecipeCache: subRecipeCache,
        recipeName: recipe.name,
        visited: {},
        onIngredient: (ingredientId, grams, recipeName) {
          final byRecipe =
              aggregated.putIfAbsent(ingredientId, () => {});
          byRecipe[recipeName] = (byRecipe[recipeName] ?? 0.0) + grams;
        },
      );
    }

    // 7. Build ShoppingItems with per-recipe contributions.
    final items = <ShoppingItem>[];
    for (final entry in aggregated.entries) {
      final ing = ingredientCache[entry.key];
      if (ing == null) continue;
      final contributions = entry.value.entries
          .map((e) => RecipeContribution(recipeName: e.key, grams: e.value))
          .toList()
        ..sort((a, b) => a.recipeName.compareTo(b.recipeName));
      items.add(ShoppingItem(
        ingredientId: entry.key,
        name: ing.name,
        totalGrams: contributions.fold(0.0, (s, c) => s + c.grams),
        unit: ShoppingUnit.grams,
        contributions: contributions,
      ));
    }

    // 8. Group by category.
    final categoryMap = <IngredientCategory?, List<ShoppingItem>>{};
    for (final item in items) {
      final cat = ingredientCache[item.ingredientId]?.category;
      categoryMap.putIfAbsent(cat, () => []).add(item);
    }

    final categories = categoryMap.entries
        .map((e) => ShoppingCategory(
              name: e.key?.displayName ?? 'Other',
              emoji: e.key?.emoji ?? '🛒',
              items: e.value..sort((a, b) => a.name.compareTo(b.name)),
            ))
        .toList()
      ..sort((a, b) => a.name.compareTo(b.name));

    return ShoppingList(
      period: ShoppingPeriod(
        startDate: startDate,
        endDate: endDate,
        recipes: allRecipes
            .map((r) => SelectableRecipe(
                  recipeId: r.id,
                  recipeName: r.name,
                ))
            .toList()
          ..sort((a, b) => a.recipeName.compareTo(b.recipeName)),
      ),
      categories: categories,
    );
  }

  // ── Helpers ───────────────────────────────────────────────────────────────

  Set<String> _collectAllIngredientIds(
    List<Recipe> recipes,
    Map<String, Recipe> subRecipeCache,
  ) {
    final ids = <String>{};
    for (final recipe in recipes) {
      for (final ri in recipe.ingredients) {
        if (ri.ingredientId != null) {
          ids.add(ri.ingredientId!);
        } else if (ri.subRecipeId != null) {
          final sub = subRecipeCache[ri.subRecipeId!];
          if (sub != null) {
            ids.addAll(_collectAllIngredientIds([sub], subRecipeCache));
          }
        }
      }
    }
    return ids;
  }

  void _expandRecipeIngredients({
    required Recipe recipe,
    required double portionMultiplier,
    required Map<String, Recipe> subRecipeCache,
    required String recipeName,
    required Set<String> visited,
    required void Function(
            String ingredientId, double grams, String recipeName)
        onIngredient,
  }) {
    if (visited.contains(recipe.id)) return;
    visited.add(recipe.id);

    for (final ri in recipe.ingredients) {
      if (ri.ingredientId != null && ri.grams != null) {
        onIngredient(
            ri.ingredientId!, ri.grams! * portionMultiplier, recipeName);
      } else if (ri.subRecipeId != null && ri.portions != null) {
        final sub = subRecipeCache[ri.subRecipeId!];
        if (sub == null) continue;
        _expandRecipeIngredients(
          recipe: sub,
          portionMultiplier: portionMultiplier * ri.portions!,
          subRecipeCache: subRecipeCache,
          recipeName: recipeName,
          visited: visited,
          onIngredient: onIngredient,
        );
      }
    }
  }
}

// ── Simple math expression evaluator ─────────────────────────────────────────
//
// Supports +, -, *, / and parentheses.  Returns null if the expression is
// invalid.  Used for the shopping list quantity adjustment feature.

double? evaluateExpression(String expr) {
  try {
    final result = _Parser(expr.trim()).parse();
    if (result.isNaN || result.isInfinite) return null;
    return result;
  } catch (_) {
    return null;
  }
}

class _Parser {
  _Parser(this._input);

  final String _input;
  int _pos = 0;

  double parse() {
    final result = _parseAddSub();
    _skipWs();
    if (_pos < _input.length) throw FormatException('Unexpected char');
    return result;
  }

  double _parseAddSub() {
    var left = _parseMulDiv();
    _skipWs();
    while (_pos < _input.length &&
        (_input[_pos] == '+' || _input[_pos] == '-')) {
      final op = _input[_pos++];
      final right = _parseMulDiv();
      left = op == '+' ? left + right : left - right;
      _skipWs();
    }
    return left;
  }

  double _parseMulDiv() {
    var left = _parseUnary();
    _skipWs();
    while (_pos < _input.length &&
        (_input[_pos] == '*' || _input[_pos] == '/')) {
      final op = _input[_pos++];
      final right = _parseUnary();
      left = op == '*' ? left * right : left / right;
      _skipWs();
    }
    return left;
  }

  double _parseUnary() {
    _skipWs();
    if (_pos < _input.length && _input[_pos] == '-') {
      _pos++;
      return -_parsePrimary();
    }
    if (_pos < _input.length && _input[_pos] == '+') {
      _pos++;
    }
    return _parsePrimary();
  }

  double _parsePrimary() {
    _skipWs();
    if (_pos < _input.length && _input[_pos] == '(') {
      _pos++; // consume '('
      final result = _parseAddSub();
      _skipWs();
      if (_pos >= _input.length || _input[_pos] != ')') {
        throw FormatException('Expected )');
      }
      _pos++; // consume ')'
      return result;
    }
    return _parseNumber();
  }

  double _parseNumber() {
    _skipWs();
    final start = _pos;
    while (_pos < _input.length &&
        (_input[_pos] == '.' ||
            (_input[_pos].codeUnitAt(0) >= 48 &&
                _input[_pos].codeUnitAt(0) <= 57))) {
      _pos++;
    }
    if (_pos == start) throw FormatException('Expected number at pos $_pos');
    return double.parse(_input.substring(start, _pos));
  }

  void _skipWs() {
    while (_pos < _input.length && _input[_pos] == ' ') {
      _pos++;
    }
  }
}
